sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala

```diff
@@ -305,6 +262,90 @@ class BulkWriter(container: CosmosAsyncContainer,
+    isGettingRetried: AtomicBoolean,
+    responseException: Option[CosmosException]
+  ) : Unit = {
+
+    val exceptionMessage = responseException match {
+      case Some(e) => e.getMessage
+      case None => ""
+    }
+
+    log.logDebug(s"encountered item operation response with status code " +
+      s"${itemResponse.getStatusCode}:${itemResponse.getSubStatusCode}, " +
+      s"Context: ${operationContext.toString} ${getThreadInfo}")
+    if (shouldIgnore(itemResponse.getStatusCode, itemResponse.getSubStatusCode, context)) {
+      log.logDebug(s"for itemId=[${context.itemId}], partitionKeyValue=[${context.partitionKeyValue}], " +
+        s"ignored encountered status code '${itemResponse.getStatusCode}:${itemResponse.getSubStatusCode}', " +
+        s"Context: ${operationContext.toString}")
+      totalSuccessfulIngestionMetrics.getAndIncrement()
+      // work done
+    } else if (shouldRetry(itemResponse.getStatusCode, itemResponse.getSubStatusCode, context)) {
+      // requeue
+      log.logWarning(s"for itemId=[${context.itemId}], partitionKeyValue=[${context.partitionKeyValue}], " +
+        s"encountered status code '${itemResponse.getStatusCode}:${itemResponse.getSubStatusCode}', will retry! " +
+        s"attemptNumber=${context.attemptNumber}, exceptionMessage=${exceptionMessage},  " +
+        s"Context: {${operationContext.toString}} ${getThreadInfo}")
+
+      // this is to ensure the submission will happen on a different thread in background
+      // and doesn't block the active thread
+      val deferredRetryMono = SMono.defer(() => {
+        scheduleWriteInternal(itemOperation.getPartitionKeyValue,
+          itemOperation.getItem.asInstanceOf[ObjectNode],
+          OperationContext(context.itemId, context.partitionKeyValue, context.eTag, context.attemptNumber + 1))
+        SMono.empty
+      })
+
+      if (Exceptions.isTimeout(itemResponse.getStatusCode)) {
+        deferredRetryMono
+          .delaySubscription(
+            Duration(
+              BulkWriter.minDelayOn408RequestTimeoutInMs +
+                scala.util.Random.nextInt(
+                  BulkWriter.maxDelayOn408RequestTimeoutInMs - BulkWriter.minDelayOn408RequestTimeoutInMs),
+              TimeUnit.MILLISECONDS),
+            Schedulers.boundedElastic())
+          .subscribeOn(Schedulers.boundedElastic())
+          .subscribe()
+
+      } else {
+        deferredRetryMono
+          .subscribeOn(Schedulers.boundedElastic())
+          .subscribe()
+      }
+
+      isGettingRetried.set(true)
+    } else {
+      val payload = SparkUtils.objectNodeToJson(itemOperation.getItem.asInstanceOf[ObjectNode])
+
+      log.logError(s"for itemId=[${context.itemId}], partitionKeyValue=[${context.partitionKeyValue}], " +
+        s"encountered status code '${itemResponse.getStatusCode}:${itemResponse.getSubStatusCode}', all retries exhausted! " +
+        s"attemptNumber=${context.attemptNumber}, exceptionMessage=${exceptionMessage}, payload=$payload, " +
```

Comment:
do we have any security concern to log the whole payload?

Reply:
Valid question - I removed it from the exception message for the same reason. IMO keeping it in logs shouldn't be a security issue because we wouldn't log it on our side - but it is definitely a gray area - especially when customers use end-to-end encryption (we don't support it yet in Spark but will need to be added at some point) raw payloads shouldn't even be logged in the Spark environment (could also be Synapse after all). So, I agree - I will remove it. Good catch!